_blog/2024-02-08-uptime-monitoring-rust.mdx

```diff
@@ -0,0 +1,694 @@
+            alias: website.alias,
+            data,
+        });
+    }
+
+    Ok(WebsiteLogs { logs })
+```
+
+Once this is done, we’ll then need to make a new Vector. This will hold all the website URLs (and respective aliases) as well as a list of timestamps with the uptime percentage over the last 24 hours, calculated per hour from how many HTTP requests returned with `200 OK`. We then need to check if there’s any gaps and fill them in with a `None`. This lets the person viewing the data know that no data was recorded at the time (for example if we’re either developing locally, or if the service itself had an outage and was unable to record data). 
+
+To start with, let’s write a function for getting the daily stats for a given URL:
+
+```rust
+#[derive(sqlx::FromRow, Serialize)]
+pub struct WebsiteStats {
+    time: DateTime<Utc>,
+    uptime_pct: Option<i16>,
+}
+
+async fn get_daily_stats(alias: &str, db: &PgPool) -> Result<Vec<WebsiteStats>, ApiError> {
+    let data = sqlx
+        ::query_as::<_, WebsiteStats>(
+            r#"
+            SELECT date_trunc('hour', created_at) as time,
+            CAST(COUNT(case when status = 200 then 1 end) * 100 / COUNT(*) AS int2) as uptime_pct
+            FROM logs
+            LEFT JOIN websites on websites.id = logs.website_id
+            WHERE websites.alias = $1
+            group by time
+            order by time asc
+            limit 24
+            "#
+        )
+        .bind(alias)
+        .fetch_all(db).await?;
+
+    let number_of_splits = 24;
+    let number_of_seconds = 3600;
+
+    let data = fill_data_gaps(data, number_of_splits, SplitBy::Hour, number_of_seconds);
+
+    Ok(data)
+}
+```
+
+Although the SQL function looks quite complicated, it basically retrieves a truncated date (down to the hour) and an uptime percentage based on the recorded data and what percentage recorded `200 OK` response status. The query then aggregates them by timestamp and limits it to the last 24 hours.
+
+Of course, this can lead to some awkward gaps in our data - for example, what if our monitoring web service goes down? We can’t exactly just go back in time and record the data! We can, however, make up for this by filling in the gaps with a new `fill_data_gaps` function that will inform the webpage viewer that there’s no data points for a given time.
+
+We can do this by declaring an enum:
+
+```rust
+enum SplitBy {
+    Hour,
+    Day
+}
+```
+
+This enum will allow us to differentiate what period we want data over. We can abstract this block into its own function:
+
+```rust
+fn fill_data_gaps(
+    mut data: Vec<WebsiteStats>,
+    splits: i32,
+    format: SplitBy,
+    number_of_seconds: i32
+) -> Vec<WebsiteStats> {
+    // if the length of data is not as long as the number of required splits (24)
+    // then we fill in the gaps
+    if (data.len() as i32) < splits {
+        // for each split, format the time and check if the timestamp exists
+        for i in 1..24 {
```

Comment:
I'm pretty sure there is a bug here, but need to spend time thinking about it...